### PR TITLE
fix: Better way of walking types in `TypeExpression`

### PR DESCRIPTION
**Pull request by mvorisek**
Previously the indexes were not updated at all.

`array_reverse` was more like a hack, the walk order should be not reversed. We should simply do DFS walk.

### CODE REVIEW

File: src/DocBlock/TypeExpression.php [hunk 1]
```diff
@@ -271,20 +271,26 @@ public function getTypesGlue(): string
...
-        foreach (array_reverse($this->innerTypeExpressions) as [
-            'start_index' => $startIndex,
+        $startIndexOffset = 0;
+
+        foreach ($this->innerTypeExpressions as $k => [
+            'start_index' => $startIndexOrig,
             'expression' => $inner,
         ]) {
-            $initialValueLength = \strlen($inner->toString());
+            $this->innerTypeExpressions[$k]['start_index'] += $startIndexOffset; // @phpstan-ignore-line https://github.com/phpstan/phpstan/issues/11171
```
Comment (mvorisek) [inline src/DocBlock/TypeExpression.php:280]: How to fix the PHPStan issue to satisfy the CS rules?

**Comment by mvorisek** [comment]
PR is done and CI is happy now.

**Comment by mvorisek** [comment]
The PR is done, we cannot omit inner reparse if the expression value has changed.

**Comment by mvorisek** [comment]
@Wirone can I please have your review here so I can work on following PR?

**Comment by mvorisek** [comment]
@Wirone can I please ask you to review this PR, I need it for #7619. Thank youl

**Comment by mvorisek** [comment]
@Wirone there is something wrong with the parallel runner and MacOS, see one CI output:
```
Errors output:

1) PhpCsFixer\Tests\Runner\RunnerTest::testParallelFixStopsOnFirstViolationIfSuchOptionIsEnabled with data set "stop on violation"
Unknown process identifier: php-cs-fixer_parallel_6668d90755ed13.75106657

/Users/runner/work/PHP-CS-Fixer/PHP-CS-Fixer/src/Runner/Parallel/ParallelisationException.php:28
/Users/runner/work/PHP-CS-Fixer/PHP-CS-Fixer/src/Runner/Parallel/ProcessPool.php:49
/Users/runner/work/PHP-CS-Fixer/PHP-CS-Fixer/src/Runner/Runner.php:216
/Users/runner/work/PHP-CS-Fixer/PHP-CS-Fixer/vendor/evenement/evenement/src/EventEmitterTrait.php:143
/Users/runner/work/PHP-CS-Fixer/PHP-CS-Fixer/vendor/clue/ndjson-react/src/Decoder.php:139
/Users/runner/work/PHP-CS-Fixer/PHP-CS-Fixer/vendor/evenement/evenement/src/EventEmitterTrait.php:143
/Users/runner/work/PHP-CS-Fixer/PHP-CS-Fixer/vendor/react/stream/src/Util.php:71
/Users/runner/work/PHP-CS-Fixer/PHP-CS-Fixer/vendor/evenement/evenement/src/EventEmitterTrait.php:143
/Users/runner/work/PHP-CS-Fixer/PHP-CS-Fixer/vendor/react/stream/src/DuplexResourceStream.php:209
/Users/runner/work/PHP-CS-Fixer/PHP-CS-Fixer/vendor/react/event-loop/src/StreamSelectLoop.php:246
/Users/runner/work/PHP-CS-Fixer/PHP-CS-Fixer/vendor/react/event-loop/src/StreamSelectLoop.php:213
/Users/runner/work/PHP-CS-Fixer/PHP-CS-Fixer/src/Runner/Runner.php:349
/Users/runner/work/PHP-CS-Fixer/PHP-CS-Fixer/src/Runner/Runner.php:157
/Users/runner/work/PHP-CS-Fixer/PHP-CS-Fixer/tests/Runner/RunnerTest.php:218
```

you should be able to reproduce it with enough pushes, possibly using a CI workflow with many `CI / PHP 8.3 tests on macOS` tests

**Comment by mvorisek** [comment]
@keradus can you please review?

**Comment by mvorisek** [comment]
@Wirone can this bugfix be merged please?

**Comment by mvorisek** [comment]
Dear maintainers, can this PR be reviewed?

**Comment by mvorisek** [comment]
@keradus can you please review?

File: src/DocBlock/TypeExpression.php [hunk 2]
```diff
@@ -393,7 +407,9 @@ private function parse(): void
             $consumedValueLength = \strlen($matches[0][0]);
             $index += $consumedValueLength;
 
-            if (\strlen($this->value) === $index) {
+            if (\strlen($this->value) <= $index) {
+                \assert(\strlen($this->value) === $index);
+
```
Comment (Wirone) [inline src/DocBlock/TypeExpression.php:412]: Is this change actually needed? All tests pass without it. Tests aside, I don't fully get the logic behind it - entering condition for `<=` and then making an assert for `===`? If all `<` values are invalid there, why the `if` change?

File: tests/DocBlock/TypeExpressionTest.php [hunk 3]
```diff
@@ -651,20 +651,42 @@ public static function provideAllowsNullCases(): iterable
 
     public function testWalkTypes(): void
     {
-        $typeExpression = new TypeExpression('Foo|Bar|Baz', null, []);
-        $addLeadingSlash = static function (TypeExpression $type): void {
+        $typeExpression = new TypeExpression('Foo|Bar|($v is \Closure(X, Y): Z ? U : (V&W))', null, []);
+
+        $addLeadingSlashFx = static function (TypeExpression $type): void {
```
Comment (Wirone) [inline tests/DocBlock/TypeExpressionTest.php:656]: Personally I don't like the `fx` convention. What is the meaning of that?

**Review by Wirone — COMMENTED** [review]
Only small doubts from my side 🙂.

File: tests/DocBlock/TypeExpressionTest.php (hunk 3 above)
Comment (mvorisek) [reply tests/DocBlock/TypeExpressionTest.php:656]: callable variable - `$addLeadingSlash` could be otherwise read as a flag, `Fx` suffix is good coding convention for more self explanatory code

File: src/DocBlock/TypeExpression.php (hunk 2 above)
Comment (mvorisek) [reply src/DocBlock/TypeExpression.php:412]: the logic is non trivial and critical - the `assert` is for safety only to assert the loop exists and always terminates

File: tests/DocBlock/TypeExpressionTest.php (hunk 3 above)
Comment (Wirone) [reply tests/DocBlock/TypeExpressionTest.php:656]: We have multiple closures like this in the `src` and there is no such convention, and I am not sure we should introduce it. Let's make those variables compatible with rest of the codebase (drop `Fx` suffix). I am not totally against this idea, but I like standardisation, so if we want such suffix, let's enforce it and change all the occurences 🙂.

Comment (Wirone) [reply tests/DocBlock/TypeExpressionTest.php:656]: In `master` there are only to `Fx` usages, in `TypeExpressionTest::testSortTypes()`, done by you in the past. Since this is out of the scope of this PR, we can keep them, but let's not introduce 2 new occurrences.

Comment (mvorisek) [reply tests/DocBlock/TypeExpressionTest.php]: done

**Review by Wirone — APPROVED** [review]
